### PR TITLE
Keep track of out of order notifications in the LSP Progress Manager

### PR DESCRIPTION
**Pull request by rubenporras**
WorkDoneProgressNotification that arrive before the job has been
started are kept in an orphan, which are checked by the jobs so that 
we do not have orphan jobs which run forever.

### CODE REVIEW

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/progress/LSPProgressManager.java [hunk 1]
```diff
@@ -149,10 +157,15 @@ private void report(final WorkDoneProgressReport report, final IProgressMonitor
 	 *            the {@link ProgressParams} used for the progress notification
 	 */
 	public void notifyProgress(final @NonNull ProgressParams params) {
-		String jobName = params.getToken().getLeft();
-		BlockingQueue<ProgressParams> progress = progressMap.get(jobName);
+		String jobIdentifier = params.getToken().getLeft();
```
Comment (mickaelistria) [inline org.eclipse.lsp4e/src/org/eclipse/lsp4e/progress/LSPProgressManager.java:160]:
```suggestion
		String jobIdentifier = params.getToken().map(Function.identity(), Object::toString);
```
and also on line 61.

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/progress/LSPProgressManager.java [hunk 2]
```diff
@@ -63,7 +67,10 @@ public void connect(final LanguageServer languageServer) {
 			LanguageServerPlugin.logInfo(
 					"Old progress with identifier " + jobIdentifier + " discarded due to new create progress request"); //$NON-NLS-1$//$NON-NLS-2$
 		}
+		return CompletableFuture.runAsync(() -> createJob(languageServer, queue, jobIdentifier));
```
Comment (mickaelistria) [inline org.eclipse.lsp4e/src/org/eclipse/lsp4e/progress/LSPProgressManager.java:70]: I don't think calling `createJob` in a `runAsync()` is a good thing. Jobs are already async, so here we may wait for more time before actually running: the time it takes to the future to be run + the time it takes for the job to be scheduled.
As we do not need the server to wait before returning here (maybe we'll need the server to wait some day on the Job *completion* but not on the Job *creation*), it seems more efficient to call `createJob` and then return directly ` CompletableFuture.completedFuture(null)`

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/progress/LSPProgressManager.java [hunk 3]
```diff
@@ -38,10 +40,12 @@ public class LSPProgressManager {
 	private final Map<String, BlockingQueue<ProgressParams>> progressMap;
```
Comment (mickaelistria) [inline org.eclipse.lsp4e/src/org/eclipse/lsp4e/progress/LSPProgressManager.java:40]: Is it worth keeping all the params or only the last one? Is there value in processing intermediary reports when we have multiple ones to process.

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/progress/LSPProgressManager.java [hunk 4]
```diff
@@ -38,10 +40,12 @@ public class LSPProgressManager {
 	private final Map<String, BlockingQueue<ProgressParams>> progressMap;
 	private final Map<IProgressMonitor, Integer> currentPercentageMap;
 	private LanguageServer languageServer;
+	private final Set<String> orphanSet;
```
Comment (mickaelistria) [inline org.eclipse.lsp4e/src/org/eclipse/lsp4e/progress/LSPProgressManager.java:43]: What about just calling it `done` ?

**Review by mickaelistria — COMMENTED** [review]
Thanks.
I've added some unrelated command because I didn't have the time to review the previous commit properly and now have some more time to consider it. So some comments are not related to that particular change. You're free about how to deal with those, nothing is mandatory.
On a general note, I think it would be better to create a `LSPProgressJob` directly, that would store information about the progress instead of the manager. The manager would simply forward progress notifications to the right job. I believe this would make things simpler.

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/progress/LSPProgressManager.java (hunk 1 above)
Comment (rubenporras) [reply org.eclipse.lsp4e/src/org/eclipse/lsp4e/progress/LSPProgressManager.java]: That is nice, I did not know that API. Thanks.

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/progress/LSPProgressManager.java (hunk 4 above)
Comment (rubenporras) [reply org.eclipse.lsp4e/src/org/eclipse/lsp4e/progress/LSPProgressManager.java]: done

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/progress/LSPProgressManager.java (hunk 3 above)
Comment (rubenporras) [reply org.eclipse.lsp4e/src/org/eclipse/lsp4e/progress/LSPProgressManager.java:41]: I thought about keeping all messages, but in the end if the progress is still running when the job starts, the job will pick some new report soon and then the work will be reflected. Thus for the time being I would to keep it simple. Maybe later someone  (or me I have time implements that)

**Comment by rubenporras** [comment]
> On a general note, I think it would be better to create a LSPProgressJob directly, that would store information about the progress instead of the manager. The manager would simply forward progress notifications to the right job. I believe this would make things simpler.

@mickaelistria , I will consider that for another PR, for the moment I would like just to avoid endless progress bars.

**Pull request merged**